src/extension/src/TelemetryWriter.py

```diff
@@ -13,75 +13,123 @@
+        self.__delete_older_events()
+
+        event = self.__new_event_json(task_name, event_level, message)
+        if len(json.dumps(event)) > Constants.TELEMETRY_EVENT_SIZE_LIMIT_IN_BYTES:
+            print("Cannot send data to telemetry as it exceeded the acceptable data size. [Data not sent={0}]".format(json.dumps(message)))
+        else:
+            self.write_event_using_temp_file(self.events_folder_path, event)
+
+    def __delete_older_events(self):
+        """ Delete older events until the at least one new event file can be added as per the size restrictions """
+        if self.__get_events_dir_size() < Constants.TELEMETRY_DIR_SIZE_LIMIT_IN_BYTES - Constants.TELEMETRY_EVENT_FILE_SIZE_LIMIT_IN_BYTES:
+            # Not deleting any existing event files as the event directory does not exceed max limit. At least one new event file can be added. Not printing this statement as it will add repetitive logs
+            return
+
+        print("Events directory size exceeds maximum limit. Deleting older event files until at least one new event file can be added.")
+        event_files = [os.path.join(self.events_folder_path, event_file) for event_file in os.listdir(self.events_folder_path) if (event_file.lower().endswith(".json"))]
+        event_files.sort(key=os.path.getmtime, reverse=True)
+
+        for event_file in event_files:
+            try:
+                if self.__get_events_dir_size() < Constants.TELEMETRY_DIR_SIZE_LIMIT_IN_BYTES - Constants.TELEMETRY_EVENT_FILE_SIZE_LIMIT_IN_BYTES:
+                    # Not deleting any more event files as the event directory has sufficient space to add at least one new event file. Not printing this statement as it will add repetitive logs
+                    break
+
+                if os.path.exists(event_file):
+                    os.remove(event_file)
+                    print("Deleted event file. [File={0}]".format(repr(event_file)))
+            except Exception as e:
+                print("Error deleting event file. [File={0}] [Exception={1}]".format(repr(event_file), repr(e)))
+
+        if self.__get_events_dir_size() >= Constants.TELEMETRY_DIR_SIZE_LIMIT_IN_BYTES:
+            raise Exception("Older event files were not deleted. Current event will not be sent to telemetry as events directory size exceeds maximum limit")
+
+    def write_event_using_temp_file(self, folder_path, data, mode='w'):
+        """ Writes to a temp file in a single operation and then moves/overrides the original file with the temp """
+        file_path = self.__get_event_file_path(folder_path)
+        prev_events = []
+        try:
+            if os.path.exists(file_path):
+                # if file_size exceeds max limit, sleep for 1 second, so the event can be written to a new file
+                file_size = self.get_file_size(file_path)
+                if file_size >= Constants.TELEMETRY_EVENT_FILE_SIZE_LIMIT_IN_BYTES:
+                    time.sleep(1)
```

Comment:
time.sleep is a little bit suspect here.. if there's a good justification, add it as a comment in code. This question will come up for all time.sleeps I see in any PR.

Reply:
This is to ensure we don't exceed the agent limit for max # of events in an event file.

We are writing multiple events within a each event file. Since the event file name is based on the current timestamp, in case we have reached the limit for events that can exist in an event file, the handler would wait for sometime, so the next attempt to write the current event will be in a new file. Sleep is to ensure a new file is created since the file name is a timestamp (agent requirement).

Will add a comment in code

Reply:
Comment updated on line 101